### DISCUSSION TITLE
Makefile anatomy

### DISCUSSION THREAD

**Discussion by thomas-fossati** [TrustedComputingGroup/concise-evidence#1] [Q&A]
It would help me if there were more comments in the makefile explaining what each statement accomplishes. Much of it is sorta obvious but looking for confirmation that I understand it.

**Comment by thomas-fossati** [comment]
The default action, if `make` is invoked without arguments, is to execute the `check` rule:

```makefile
.DEFAULT_GOAL := check
```

The three auto-generated files assembled by the CoRIM builder (i.e., `ietf-corim-cddl/corim-autogen.cddl` & co.) are added to the `CORIM_DEPS` variable:

```makefile
CORIM_DIR := ietf-corim-cddl
CORIM_DEPS := $(foreach ca,corim comid xcorim,$(CORIM_DIR)/$(ca)-autogen.cddl)
```

The needed tools (i.e., `cddl(1)`, `cbor2diag.rb(1)`, etc.) are sourced from the CoRIM repo:

```makefile
include $(CORIM_DIR)/tools.mk
```

The `cddl_check_template`, which is used below to construct the `check-spdm` rule from all the CDDL fragments, is sourced from the CoRIM repo: 

```makefile
include $(CORIM_DIR)/funcs.mk
```

Here we tell `make` that in order to build the three CoRIM auto-generated files, it needs to call itself recursively in the `CORIM_DIR` folder:

```makefile
$(CORIM_DEPS): ; $(MAKE) -C $(CORIM_DIR)
```

We declare the `check` rule and we append the `check-spdm` goal to it:

```makefile
check:: check-spdm
```

Here we accumulate into the `SPDM_FRAGS` variable all the CDDL snippets that the spec is composed of:

```makefile
SPDM_FRAGS := spdm-toc.cddl
SPDM_FRAGS += ce-code-points.cddl
SPDM_FRAGS += comid-extensions.cddl
SPDM_FRAGS += comid-extn-code-points.cddl
SPDM_FRAGS += concise-evidence.cddl
SPDM_FRAGS += toc-code-points.cddl
SPDM_FRAGS += $(CORIM_DEPS)
```

Then we declare which examples we want to automatically check against the spec.  Note thought that the generated rule is not used at the moment because the CDDL doesn't validate yet.  We'll append it to the `check` goal later.

```makefile
SPDM_EXAMPLES := $(wildcard examples/spdm-*.diag)
```

This is where all the "check" rules are synthesised by invoking the `cddl_check_template` function and evaluating it on the supplied arguments:

```makefile
$(eval $(call cddl_check_template,spdm,$(SPDM_FRAGS),$(SPDM_EXAMPLES)))
```

This is the rule that cleans up the mess in the current folder:

```makefile
clean: ; $(RM) $(CLEANFILES)
```

This is the rule that cleans up the mess in the current folder as well as in the CoRIM sub folder:

```makefile
clean-extra: clean ; $(MAKE) -C $(CORIM_DIR) clean
```

**Reply by nedmsmith** [reply]
The SPDM_EXAMPLES target is asserting all example files should be prepended with 'spdm_'(?) even though some examples are rim / tag files that exercises extensions to corim. Should these be called 'corim_' or 'comid_' examples instead? BTW: what does wildcard accomplish?

**Reply by thomas-fossati** [reply]
> The SPDM_EXAMPLES target is asserting all example files should be prepended with 'spdm_'(?) even though some examples are rim / tag files that exercises extensions to corim. Should these be called 'corim_' or 'comid_' examples instead?

It should probably be `$(wildcard examples/*.diag)`

**Reply by nedmsmith** [reply]
Makes sense

**Comment by thomas-fossati** [comment]
> BTW: what does wildcard accomplish?

`$(wildcard ...)` expands to the list of names of existing files that match one of the given file name patterns.

**Comment by nedmsmith** [comment]
Makefile requires multiple targets for testing comid extensions separately from concise-evidence. These were added and examples based on the core schema execute cleanly. however, anything involving an extension point doesn't seem to be recognized.
